topology/src/main/java/io/camunda/zeebe/topology/gossip/ClusterTopologyGossiper.java

```diff
@@ -0,0 +1,188 @@
+
+    LOGGER.trace("Sending sync request to {}", randomMemberToSync);
+    communicationService
+        .send(
+            SYNC_REQUEST_TOPIC,
+            gossipState,
+            serializer::encode,
+            serializer::decode,
+            randomMemberToSync,
+            config.syncRequestTimeout())
+        .whenCompleteAsync(
+            (response, error) -> handleSyncResponse(response, error, randomMemberToSync),
+            executor::run);
+  }
+
+  private void refreshMembersToSync() {
+    if (membersToSync.isEmpty()) {
+      membersToSync =
+          new LinkedList<>(membershipService.getMembers().stream().map(Member::id).toList());
+      membersToSync.remove(membershipService.getLocalMember().id());
+      Collections.shuffle(membersToSync);
+    }
+  }
+
+  private void handleSyncResponse(
+      final ClusterTopologyGossipState response, final Throwable error, final MemberId member) {
+    if (error == null) {
+      update(response);
+    } else {
+      LOGGER.debug("Failed to sync with {}", member, error);
+    }
+    scheduleSync();
+  }
+
+  // returns true if local state is changed
+  private boolean update(final ClusterTopologyGossipState response) {
+    if (!response.equals(gossipState)) {
+      final ClusterTopology topology = response.getClusterTopology();
+      if (topology != null) {
+        final var updatedTopology = clusterTopologyUpdateHandler.apply(topology);
+        gossipState.setClusterTopology(updatedTopology);
+        LOGGER.trace("Updated local gossipState to {}", updatedTopology);
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  private ClusterTopologyGossipState handleSyncRequest(
+      final MemberId memberId, final ClusterTopologyGossipState clusterSharedGossipState) {
+    LOGGER.trace(
+        "Received topology sync request from {} with state {}", memberId, clusterSharedGossipState);
+    update(clusterSharedGossipState);
+    return gossipState;
+  }
+
+  public void updateClusterTopology(final ClusterTopology clusterTopology) {
+    if (clusterTopology == null) {
+      return;
+    }
+    executor.run(
+        () -> {
+          if (!clusterTopology.equals(gossipState.getClusterTopology())) {
+            gossipState.setClusterTopology(clusterTopology);
+            gossip();
+          }
+        });
+  }
+
+  private void gossip() {
+    // TODO: Instead of selecting random members, we can also propagate via a tree topology to
+    // prevent duplicate gossip updates
+    refreshMembersToSync();
+    if (membersToSync.isEmpty()) {
+      return;
+    }
+    final var gossipMembersList =
+        membersToSync.subList(0, Math.min(config.gossipFanout(), membersToSync.size()));
+    LOGGER.trace("Gossiping {} to {}", gossipState, gossipMembersList);
+    gossipMembersList.forEach(
+        member ->
+            communicationService.unicast(
+                SYNC_REQUEST_TOPIC, gossipState, serializer::encode, member, true));
```

Comment:
:question: SWIM gossips over UDP. But here is TCP. Should we switch back to UDP?

Reply:
imo using UDP with our current stack is probably wrong =/ Since UDP doesn't guarantee ordering, anything which gets split up into multiple packets may end up scrambled on the other side, and I don't think we check for that. It's probably low risk in our cloud, but :shrug:

Reply:
:+1: I will keep it as TCP.